harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+        }
+
+        for path in self._paths:
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        return metrics
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def collect(self) -> Dict[str, Any]:
+        free_mem_bytes = psutil.virtual_memory().available
+        return {
+            "free_memory": free_mem_bytes / 1e9,
+        }
+
+
+class _CPU(_MetricGroupCollector):
+    group = "cpu"
+
+    def collect(self) -> Dict[str, Any]:
+        cpu_util = psutil.cpu_percent()
+        return {
+            "cpu_util_simple": cpu_util,
+        }
+
+
+class _GPU(_MetricGroupCollector):
+    """
+    Collects GPU metrics for all GPU devices on the machine.
+
+    Returns `gpu_util` and `gpu_free_memory` metrics labeled by the device's UUID. Example:
+
+    .. code::
+        {
+            "GPU-UUID-A": {
+                "gpu_util": 0.12,
+                "gpu_free_memory": 5.662642176,
+            },
+            "GPU-UUID-B": {
+                "gpu_util": 0.23,
+                "gpu_free_memory": 6.552642176,
+            },
+        }
+    """
+
+    group = "gpu"
+
+    def __init__(self):
+        super().__init__()
+
+        self._pynvml_device_handles: Dict[str, Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning(f"pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        try:
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                uuid = pynvml.nvmlDeviceGetUUID(handle)
+                pynvml.nvmlDeviceGetMemoryInfo(handle)
+                pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml_device_handles[uuid] = handle
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
+            logging.error(f"Error initializing pynvml: {e}")
+
+    def collect(self) -> Dict[str, Any]:
+        metrics = {}
+
+        for uuid, handle in self._pynvml_device_handles.items():
+            gpu_util = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
+            free_memory = pynvml.nvmlDeviceGetMemoryInfo(handle).free
+            metrics.update(
+                {
+                    uuid: {
+                        "gpu_util": gpu_util,
+                        "gpu_free_memory": free_memory,
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: Dict[str, Any],
+        timestamp: datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _Collector(threading.Thread):
```

Comment:
It's taken me a while to understand that "collect" in this file means "take a reading". My expectation had been that it meant "combine a collection into a variable" (be that variable a scalar or a collection).

If you like the word "sample" instead of "collect", cool cool cool, we're on the same wavelength. But if you like "collect" better, ain't nothing wrong with that, it's a fine word to use here. Just in that case please say something about it in a docstring.

Reply:
`Collect` becomes slightly clearer with a docstring that uses the word "sample".

Reply:
I like it.